README.md
<!--
 * @Author: mrrs878@foxmail.com
 * @Date: 2021-08-18 13:58:53
 * @LastEditors: mrrs878@foxmail.com
 * @LastEditTime: 2021-08-18 16:21:59
 * @FilePath: \bytemd\README.md
-->
packages/hooks/README.md

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.2.2](https://github.com/mrrs878/gear/compare/v0.2.1...v0.2.2) (2021-08-19)


### Bug Fixes

* 修复@mrrs878/hooks中react依赖问题 ([6b65fb8](https://github.com/mrrs878/gear/commit/6b65fb8ce1f24b0c6f34f3eeb0809fe1cca1bb23))





## [0.2.1](https://github.com/mrrs878/gear/compare/v0.2.0...v0.2.1) (2021-08-19)


### Performance Improvements

* @mrrs878/sliding-puzzle添加position返回值 ([c7e8e17](https://github.com/mrrs878/gear/commit/c7e8e17f063df9e2682cf799f87b71e162261253))





# [0.2.0](https://github.com/mrrs878/gear/compare/v0.1.5...v0.2.0) (2021-08-19)


### Features

* 添加@mrrs878/drag-verify ([8a7e32f](https://github.com/mrrs878/gear/commit/8a7e32f7d4aa0191fa54e05fc1fa82e88fedcd3f))





## [0.1.5](https://github.com/mrrs878/gear/compare/v0.1.4...v0.1.5) (2021-08-19)


### Bug Fixes

* 修改hooks中react依赖为peerDependencies ([75a2624](https://github.com/mrrs878/gear/commit/75a2624959e0994ed418e795e6e88a9c71d6e163))





## [0.1.4](https://github.com/mrrs878/gear/compare/v0.1.3...v0.1.4) (2021-08-18)


### Bug Fixes

* 修改hooks/index.ts导出方式 ([07c0155](https://github.com/mrrs878/gear/commit/07c0155ded6e0bedcb59856ce2d597f9bd36559a))





## 0.1.3 (2021-08-18)


### Bug Fixes

* 添加cross-env ([40a232b](https://github.com/mrrs878/gear/commit/40a232b1f019e34fa55900ddae01280d53b6bbff))





## 0.1.2 (2021-08-18)

**Note:** Version bump only for package gear


.eslintrc.js
/*
 * @Author: mrrs878@foxmail.com
 * @Date: 2021-08-20 11:26:26
 * @LastEditors: mrrs878@foxmail.com
 * @LastEditTime: 2021-08-24 16:06:29
 * @FilePath: \gear\.eslintrc.js
 */
module.exports = {
  extends: [
    'airbnb-typescript',
    'airbnb/hooks',
  ],
  parserOptions: {
    project: './tsconfig.json',
    createDefaultProgram: true,
  },
  ignorePatterns: ['**/dist/*', '**/lib/*'],
  rules: {
    'linebreak-style': ['off', 'window'],
    'no-console': 'off',
    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off',
    'no-underscore-dangle': ['error', { allow: ['_id'] }],
    'react-hooks/exhaustive-deps': process.env.NODE_ENV === 'production' ? 'error' : 'warn',
    '@typescript-eslint/no-unused-vars': process.env.NODE_ENV === 'production' ? 'error' : 'warn',
    'react/destructuring-assignment': [0],
    'import/prefer-default-export': 0,
    'import/no-extraneous-dependencies': 0,
  },
};


rollup.config.js
// @ts-check
import fs from 'fs-extra';
import path from 'path';
import _ from 'lodash';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import json from '@rollup/plugin-json';
import babel from '@rollup/plugin-babel';
import { terser } from 'rollup-plugin-terser';
import replace from '@rollup/plugin-replace';
import postcss from 'rollup-plugin-postcss';

const production = !process.env.ROLLUP_WATCH;
const umd = process.env.UMD;

const packages = fs.readdirSync(path.resolve(__dirname, 'packages'));

const configs = packages
  .map((key) => {
    const pkg = fs.readJsonSync(`./packages/${key}/package.json`);
    if (pkg.private) return [];

    const inputFile = path.resolve('packages', key, 'lib/index.js');
    const umdName = key.startsWith('plugin-')
      ? _.camelCase(`mrrs878-${key}`)
      : 'mrrs878';

    /** @type {import('rollup').RollupOptions} */
    const common = {
      input: inputFile,
      plugins: [
        commonjs(),
        resolve({
          browser: true,
          preferBuiltins: false,
        }),
        json(),
        replace({
          preventAssignment: true, // fix warning
          'process.env.NODE_ENV': JSON.stringify(
            production ? 'production' : 'development',
          ),
        }),
        postcss({
          extract: true,
        }),
      ],
      watch: {
        clearScreen: false,
      },
    };

    /** @type {import('rollup').RollupOptions} */
    const config = {
      ...common,
      output: [
        {
          format: 'es',
          sourcemap: true,
          file: path.resolve('packages', key, pkg.module),
        },
        {
          format: 'cjs',
          sourcemap: true,
          file: path.resolve('packages', key, pkg.main),
          exports: 'auto', // fix warning
        },
      ],
      external: [
        'hast-util-sanitize/lib/github.json',
        ...Object.keys(pkg.dependencies || {}),
        ...Object.keys(pkg.peerDependencies || {}),
      ],
    };

    /** @type {import('rollup').OutputOptions} */
    const umdOutputOption = {
      format: 'umd',
      name: umdName,
      sourcemap: true,
      globals: { react: 'react' },
      inlineDynamicImports: true,
    };

    /** @type {import('rollup').RollupOptions} */
    const umdConfig = {
      ...common,
      output: [
        {
          ...umdOutputOption,
          file: path.resolve('packages', key, 'dist/index.js'),
        },
        {
          ...umdOutputOption,
          file: path.resolve('packages', key, 'dist/index.min.js'),
          plugins: [terser()],
        },
      ],
      external: Object.keys(pkg.peerDependencies || {}),
    };

    /** @type {import('rollup').RollupOptions} */
    const es5Config = {
      ...common,
      input: path.resolve('packages', key, 'lib/index.js'),
      output: [
        {
          ...umdOutputOption,
          file: path.resolve('packages', key, 'dist/index.es5.js'),
        },
        {
          ...umdOutputOption,
          file: path.resolve('packages', key, 'dist/index.es5.min.js'),
          plugins: [terser()],
        },
      ],
      plugins: [
        ...common.plugins,
        babel({
          babelHelpers: 'runtime',
          extensions: ['.js', '.mjs', '.html'],
        }),
      ],
    };

    if (umd) {
      return [config, umdConfig, es5Config];
    }
    return [config];
  })
  .flat();

/** @type {import('rollup').RollupOptions} */
const styleCommon = {
  input: 'packages/sliding-puzzle/src/dom/index.css',
  output: {
    file: 'style.js', // We don't need this file
  },
};

const styleConfigs = [
  {
    ...styleCommon,
    plugins: [
      postcss({
        extract: path.resolve(__dirname, 'packages/sliding-puzzle/dist/index.css'),
        minimize: production,
      }),
    ],
  },
];

export default [...styleConfigs, ...configs];


scripts/postinstall.mjs
/* eslint-disable import/no-extraneous-dependencies */
// @ts-check
import fs from 'fs-extra';
import path from 'path';
import mustache from 'mustache';
import _ from 'lodash';
// @ts-ignore
import { rootDir } from './utils.mjs';

function readFileSyncSafe(p) {
  try {
    return fs.readFileSync(p, 'utf8');
  } catch (err) {
    return '';
  }
}

const packagesDir = path.join(rootDir, 'packages');
const packages = fs.readdirSync(packagesDir);
const plugins = packages.filter(
  (x) => x.startsWith('plugin-') && !x.includes('-transform'),
);

// tsconfig root
fs.writeJsonSync(
  path.resolve(rootDir, 'tsconfig.json'),
  {
    files: [],
    compilerOptions: {
      jsx: 'react',
    },
    references: packages.map((p) => ({ path: `packages/${p}` })),
  },
  { spaces: 2 },
);

packages.forEach((p) => {
  // tsconfig
  const tsconfig = {
    extends: '../../tsconfig-base.json',
    include: ['src', 'src/**/*.json'], // https://github.com/microsoft/TypeScript/issues/25636#issuecomment-627111031
    compilerOptions: {
      composite: true,
      rootDir: 'src',
      outDir: 'lib',
    },
  };
  // if (p !== 'mrrs78') {
  //   tsconfig.references = [{ path: '../mrrs78' }]
  // }

  fs.writeJsonSync(path.join(packagesDir, p, 'tsconfig.json'), tsconfig, {
    spaces: 2,
  });

  // license
  fs.copyFileSync(
    path.join(rootDir, 'LICENSE'),
    path.join(packagesDir, p, 'LICENSE'),
  );

  // package.json
  const pkgPath = path.join(packagesDir, p, 'package.json');
  const pkg = fs.readJsonSync(pkgPath);
  pkg.repository = {
    type: 'git',
    url: 'https://github.com/mrrs878/gear.git',
    directory: `packages/${p}`,
  };
  pkg.main = 'dist/index.cjs.js';
  pkg.module = 'dist/index.esm.js';
  pkg.types = 'lib/index.d.ts';
  pkg.unpkg = 'dist/index.min.js';
  pkg.jsdelivr = 'dist/index.min.js';
  pkg.files = ['dist', 'lib'];
  pkg.homepage = 'https://github.com/mrrs878/gear#readme';
  fs.writeJsonSync(pkgPath, pkg, { spaces: 2 });
});

// plugins readme
plugins.forEach((p) => {
  const name = p.split('-').slice(1).join('-');
  const result = mustache.render(
    readFileSyncSafe(path.join(rootDir, 'scripts/plugin-template.md')),
    {
      name,
      importedName: _.camelCase(name.replace('-ssr', '')),
      desc: fs.readJsonSync(path.join(packagesDir, p, 'package.json'))
        .description,
    },
  );
  fs.writeFileSync(path.join(packagesDir, p, 'README.md'), result);
});

// gear readme
const readme = readFileSyncSafe(path.join(rootDir, 'README.md')).replace(
  /## Plugins\s+([\w\W])*?\s+##/,
  () => {
    const content = plugins
      .map((p) => {
        const pkg = fs.readJsonSync(path.join(packagesDir, p, 'package.json'));
        if (pkg.private) return;

        const name = p.split('-').slice(1).join('-');
        const badge = `[![npm](https://img.shields.io/npm/v/@gear/plugin-${name}.svg)](https://npm.im/@gear/plugin-${name})`
          + ' '
          + `[![gzip size](https://img.badgesize.io/https://unpkg.com/@gear/plugin-${name}/dist/index.min.js?compression=gzip)](https://unpkg.com/@gear/plugin-${name})`;
        const desc = _.upperFirst(
          pkg.description.replace('ByteMD plugin to ', ''),
        );
        // eslint-disable-next-line consistent-return
        return `| [@gear/plugin-${name}](./packages/plugin-${name}) | ${badge} | ${desc} |`;
      })
      .filter((x) => x)
      .join('\n');

    return `## Plugins

| Package | Status | Description |
| --- | --- | --- |
${content}

##`;
  },
);

if (!readFileSyncSafe(path.join(rootDir, 'README.md'))) {
  fs.writeFileSync(path.join(rootDir, 'README.md'), readme);
}

// locales
let importCode = '';
const exportObject = {};
packages.forEach((p) => {
  const localeDir = path.join(packagesDir, p, 'src/locales');
  if (fs.existsSync(localeDir) && fs.lstatSync(localeDir).isDirectory()) {
    const locales = fs.readdirSync(localeDir).map((x) => x.replace('.json', ''));
    const libName = p.startsWith('plugin') ? `@mrrs878/${p}` : p;

    locales.forEach((l) => {
      if (!exportObject[l]) exportObject[l] = {};

      const varName = _.snakeCase(`${l}-${p}`);

      importCode += `import ${varName} from '${libName}/lib/locales/${l}';\n`;
      exportObject[l][_.snakeCase(p)] = varName;
    });
  }
});


scripts/process.mjs
/*
 * @Author: mrrs878@foxmail.com
 * @Date: 2021-08-18 13:58:53
 * @LastEditors: mrrs878@foxmail.com
 * @LastEditTime: 2021-08-18 16:21:18
 * @FilePath: \bytemd\scripts\process.mjs
 */
// @ts-check
import fs from 'fs-extra'
import path from 'path'
import glob from 'glob'
import chokidar from 'chokidar'
// @ts-ignore
import { rootDir } from './utils.mjs'

// svelte file: compile
// other files: copy
async function transform(file, dir = 'lib') {
  const dest = file.replace('/src/', `/${dir}/`)
  await fs.ensureDir(path.dirname(dest))

  if (fs.statSync(file).isDirectory()) return

  await fs.copyFile(file, dest)
}

function watch(pattern, dest) {
  chokidar.watch(pattern).on('all', (event, file) => {
    console.log(event, file)
    transform(file, dest)
  })
}

function build(pattern, dir) {
  glob(pattern, (err, files) => {
    if (err) throw err

    for (let file of files) {
      transform(file, dir)
    }
  })
}

const pattern = path.resolve(
  rootDir,
  `packages/*/src/**/*.{svelte,vue,json,wxml,wxss}`
)
const mpPattern = path.resolve(
  rootDir,
  `packages/mp/src/**/*.{svelte,vue,json,wxml,wxss}`
)

if (process.argv.includes('--watch')) {
  watch(pattern)
  watch(mpPattern, 'dist')
} else {
  build(pattern)
  build(mpPattern, 'dist')
}
